src/Config.php

```diff
@@ -14,119 +14,253 @@
+                    $key
+                )
+            );
+        }
+
+        if (!is_scalar($value) && !is_array($value)) {
+            throw new \InvalidArgumentException(
+                sprintf(
+                    'Configuration can only a be scalar or array, "%s" type given instead for "%s".',
+                    gettype($value),
+                    $key
+                )
+            );
+        }
+
+        if (in_array($key, self::$protectedConfig, true)) {
+            throw new \InvalidArgumentException(
+                sprintf('Configuration key "%s" is protected and can not be overwritten.', $key)
+            );
+        }
+
+        $this->config[$type][$key] = $value;
+        $this->config[self::CONFIG_ALL][$key] = $value;
+
+        return $this;
     }
 
     /**
-     * Merges new config values with the existing ones (overriding)
+     * Merges new config values with the existing ones (overriding).
      *
      * @param array $config
      */
-    public function merge(array $config)
+    public function merge(array $config, $type)
     {
-        // override defaults with given config
         foreach ($config as $key => $val) {
-            $this->config[$key] = $val;
+            $this->set($key, $val, $type);
         }
     }
 
     /**
-     * Returns a setting
+     * Returns a config value.
      *
-     * @param string $key
+     * @param string                      $key      Single level key like "adapters" or property-path
+     *                                              "[adapters][github]"
+     * @param string                      $type     Either Config::CONFIG_SYSTEM Config::CONFIG_LOCAL
+     *                                              or Config::CONFIG_ALL
+     * @param string|int|float|bool|array $default  Default value to use when no config is found (null)
      *
-     * @return mixed
+     * @return string|int|float|bool|array
      */
-    public function get($key)
+    public function get($key, $type = self::CONFIG_ALL, $default = null)
     {
-        switch ($key) {
-            case 'cache-dir':
-                // convert foo-bar to GUSH_FOO_BAR and check if it exists since it overrides the local config
-                $env = 'GUSH_'.strtoupper(strtr($key, '-', '_'));
-
-                return rtrim(getenv($env) ?: $this->config[$key], '/\\');
-            case 'home':
-                return rtrim($this->config[$key], '/\\');
-            default:
-                $accessor = PropertyAccess::createPropertyAccessor();
-                try {
-                    return $accessor->getValue($this->config, $key);
-                } catch (NoSuchPropertyException $e) {
-                    if (!isset($this->config[$key])) {
-                        return null;
-                    }
-
-                    return $this->config[$key];
-                }
+        $this->guardConfigSlot($type);
+
+        // array_key_exists is cheaper for a single level config
```

Comment:
I replaced this with something simpler, its not `igorw/get-in` package but its based on it.
Thanks for the suggestion :+1: